inductor:enable conv+unary fusion for torch unary function

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91677
* __->__ #91609


This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor:enable conv+unary fusion for unary op is torch function

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for unary op is torch function"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for unary op is torch function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for unary op is torch function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for unary op is torch function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/mkldnn_utils.py:
```diff
@@ -0,0 +1,28 @@
+import torch
+
+# check the pattern: (nn.module, F.function) matched.
+def matches_module_function_pattern(pattern, node, modules):
```
It is a general function, not mkldnn specific? Why do we move it to mkldnn_util.py?

XiaobingSuper replied:
Yes, how about fx_utils.py?

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

desertfire commented on torch/_inductor/mkldnn.py:
```diff
@@ -510,21 +513,50 @@ def mkldnn_fuse_fx(gm: torch.fx.GraphModule, example_inputs):
+                    ):
+                        continue
+                    unary_node = create_unary_module(node)
+                    unary_node.eval()
```
Not sure if it will ever happen, but do we need to check if `node` is in `training` before calling `eval()` here?

jgong5 replied:
Good catch. I guess it is needed. A similar check was applied after the code at here too: 
https://github.com/pytorch/pytorch/pull/91609/files#diff-7c920a79e1ea5c56baef3ef14306bc2a3a4eb4a13ca7689d93e0549eccd816d4R560

XiaobingSuper replied:
For torch function, there doesn't have **training ** attr, even for **nn.module** case of that activation, the **training ** attr is not working, so I think we don't need to do such a check here.

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor:enable conv+unary fusion for torch unary function"



This PR is about to enable unary fusion which the unary is the torch function, this PR will improve timm models performance a lot.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
